File: Creational/Builder/python/BuilderPattern.py
class Person:
    def __init__(self, builder):
        """ constructor """
        self.firstName = builder.firstName
        self.lastName = builder.lastName

    def getFirstName(self):
        return self.firstName

    def getLastName(self):
        return self.lastName

    @staticmethod
    def getBuilder():
        return _Builder()

class _Builder:
    """
    Beginning underscore signifies the Builder class is private.
    """
    def firstName(self, firstName):
        self.firstName = firstName
        return self

    def lastName(self, lastName):
        self.lastName = lastName
        return self

    def build(self):
        return Person(self)

if __name__ == '__main__':
    person = Person.getBuilder().firstName("John").lastName("Doe").build()
    print(person.getFirstName() + " " + person.getLastName())

File: Creational/Builder/kotlin/BuilderPattern.kt
fun main(args: Array<String>) {
    val john = PersonBuilder {
        name = "John Smith"
        age = 20
        interests = arrayListOf("Football")
    }
    println(john.age)
}

data class Person(
        val name: String,
        val age: Int,
        val interests: ArrayList<String>
)

class PersonBuilder() {

    constructor(init: PersonBuilder.() -> Unit) : this() {
        init()
    }

    var name: String = ""
    var age: Int = 0
    var interests: ArrayList<String> = ArrayList()

    fun name(value: PersonBuilder.() -> String) = apply { name = value() }

    fun age(value: PersonBuilder.() -> Int) = apply { age = value() }

    fun interests(value: PersonBuilder.() -> ArrayList<String>) = apply { interests = value() }

    fun build() = Person(name, age, interests)

}
